src/palace/manager/service/analytics/eventdata.py

```diff
@@ -0,0 +1,197 @@
+
+
+class AnalyticsEventData(BaseModel, LoggerMixin):
+    """
+    This class represents the data that is sent to the analytics provider.
+
+    It is a Pydantic model that is used to validate the data before it is sent. It stores
+    all the data without references to any database objects, so it can be serialized, stored
+    and sent outside a database transaction.
+    """
+
+    type: str
+    start: AwareDatetime
+
+    # TODO: We include the 'end' field as a copy of 'start' because that is
+    #   what the pre-pydantic implementation did. Since this is just duplicated
+    #   we should remove it in a future release.
+    @computed_field  # type: ignore[prop-decorator]
+    @cached_property
+    def end(self) -> AwareDatetime:
+        return self.start
+
+    library_id: int
+    library_name: str
+    library_short_name: str
+    old_value: int | None
+    new_value: int | None
+
+    @computed_field  # type: ignore[prop-decorator]
+    @cached_property
+    def delta(self) -> int | None:
+        if self.new_value is None or self.old_value is None:
+            return None
+        return self.new_value - self.old_value
+
+    location: str | None
+    license_pool_id: int | None
+    publisher: str | None
+    imprint: str | None
+    issued: datetime | None
+    published: datetime | None
+    medium: str | None
+    collection: str | None
+    collection_id: int | None
+    identifier_type: str | None
+    identifier: str | None
+    data_source: str | None
+    distributor: str | None
+    audience: str | None
+    fiction: bool | None
+    summary_text: str | None
+    quality: float | None
+    rating: int | None
+    popularity: int | None
+    genre: str | None
+    availability_time: AwareDatetime | None
+    licenses_owned: int | None
+    licenses_available: int | None
+    licenses_reserved: int | None
+    patrons_in_hold_queue: int | None
+
+    # TODO: We no longer support self-hosted books, so this should always be False.
+    #  this value is still included in the response for backwards compatibility,
+    #  but should be removed in a future release.
+    self_hosted: bool = False
+    title: str | None
+    author: str | None
+    series: str | None
+    series_position: int | None
+    language: str | None
+    open_access: bool | None
+    user_agent: str | None
+    patron_uuid: UUID | None
+
+    model_config = ConfigDict(
+        frozen=True,
+    )
+
+    # We serialize the datetime fields as strings in the JSON output, to match what
+    # the output looked like before we switched to Pydantic.
+    # TODO: It would be nice to be able to drop this and just use the default
+    #   which is to serialize datetimes as ISO8601 strings. Need to see if the
+    #   analytics consumers can handle that.
+    @field_serializer("start", "end", "availability_time", when_used="json")
+    def serialize_dt(self, value: datetime, _info: FieldSerializationInfo) -> str:
+        return str(value)
```

Comment:
@dbernstein what sort of logic are we using to parse this timestamp when we are ingesting this data into redshift? I did this for compatibility with the previous implementation, but depending on that parsing, we might be able to drop this custom serializer.

Reply:
We are not doing anything special.  We pull the dates as strings (presumably iso8601)  into a temporary database and then they are automatically converted to `timestamp` type in redshift.

Reply:
Okay. I'll remove this as well!